GH-1424 Added Jenskinsfile for PR triggered build at HAW Hamburg

jpilgrim opened the pull request:
This PR adds a jenkinsfile to the root folder. When this file is found, a build job is triggered running all tests on a pre-merged version of N4JS to ensure that the PR does not break anything.

(Oh.. I maybe need to filter out some events such as editing the PR name... Anyway, this does not effect the N4JS sources.)

jpilgrim commented:
@mor-n4 The Jenkinsfile and my server are now working :-) Please have a look.

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+                                '-Dtycho.localArtifacts=ignore',
+                                // for adjusting the script only, renable later on:
+                                 '-Dmaven.test.failure.ignore',
+                                 "--fail-at-end",
```
Why did you introduced this?

jpilgrim replied:
Otherwise maven stopped at the first failing test

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+                                '--update-snapshots',
+                                '--show-version',
+                                '-Dtycho.localArtifacts=ignore',
+                                // for adjusting the script only, renable later on:
```
Is this comment still valid?

jpilgrim replied:
I did not remove it (although I only required it for testing the script and the server) as this might be useful when updating the script or creating a new server.

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+                            def profiles = [
+                                'buildProduct',
+                                'execute-plugin-tests',
+                               // 'execute-plugin-ui-tests',
```
Why exclude the plugin-ui tests?

jpilgrim replied:
Uups... Oh no... I forgot to remove the comments here. Thanks!

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+                            // for quick testing: 7
+                            // sh "mvn -o clean"
+
+                          // sh "ls -Ral builds/org.eclipse.n4js.product.build/target/repository/"
```
Many commented-out lines.

jpilgrim replied:
yes -- they are very helpful for testing the script, this is why I did not remove them. Will add a comment

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+            steps {
+                echo 'Deploying....'
+            }
+        }*/
```
Still necessary?

jpilgrim replied:
commented out here, may be introduced again, left skeleton here -- added that comment

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+            //            """
+
+            // Execute after every other post condition has been evaluated, regardless of status
+            // See https://jenkins.io/doc/book/pipeline/syntax/#post
```
Delete? (Also last line below)

jpilgrim replied:
yes. I will adjust the clean up once the build is working stable.

mmews-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+         //   archiveArtifacts allowEmptyArchive: true, artifacts: '**/builds/**/target/products/*.zip'
+         //   archiveArtifacts allowEmptyArchive: true, artifacts: '**/tools/**/target/n4jsc.jar'
+         //   archiveArtifacts allowEmptyArchive: true, artifacts: '**/logs/*.log'
+         //   archiveArtifacts allowEmptyArchive: true, artifacts: '**/tests/**/target/**/*-output.txt'
```
Delete? (Also last line below)

jpilgrim replied:
Also here I want to maybe archive some artifacts once I really clean the build

jpilgrim commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+                echo 'Starting Xvnc'
+                wrap([$class: 'Xvnc', takeScreenshot: false, useXauthority: true]) {
+                    echo 'Building and testing, using display ' + DISPLAY
+                    // dir('n4js') {
```
removed it, required for multi-git builds only

jpilgrim commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+        YARN_PATH  = '/usr/bin' // '/shared/common/yarn/1.15.2/bin'
+        // PATH       = "${PATH}"
+                     // ":/shared/common/node-v10.15.3-linux-x64/bin"      +
+                     // ":/shared/common/java/openjdk/jdk11-x64/bin"
```
added a comment that these are common paths you may find on other machines

mor-n4 replied:
The paths with `/shared/common/...` belong to the Eclipse JIPP (Jenkins Instance Per Project) build nodes.

jpilgrim commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,108 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Based oin eclipse-nightly.jenkinsfile
```
Thanks

jpilgrim commented:
@mmews-n4 Thanks for the review. I have posted the blog (https://jevopisdeveloperblog.blogspot.com/2019/10/setup-github-triggered-build-machine.html) and added a link top that post in the design document. Also addressed most of the other comments. I will probably slightly change the build in a couple of weeks, when I got some experience (e.g. cleaning the workspace and archiving some stuff).

I will merge this PR once the new build (with enabled plugin-UI tests) is done.

mor-n4 commented on haw-pr.jenkinsfile:
```diff
@@ -0,0 +1,104 @@
+         //   archiveArtifacts allowEmptyArchive: true, artifacts: '**/tests/**/target/**/*-output.txt'
+
+            junit '**/surefire-reports/**/*.xml'
+            //junit '**/failsafe-reports/**/*.xml'
```
Having this line commented out might be a problem. If you are executing failsafe tests above and this line is commented out, the build will be marked as successful (i.e. "green") even if one or more failsafe tests did actually fail.

A short while ago, I noticed this situation in file `eclipse-nightly.jenkinsfile` and wanted to investigate further at the end of my current task, i.e. `GH-1503`.

If you want to test this now: just comment in this line and it will produce an error if no failsafe tests were running and you can safely comment it out, again. Otherwise you know this line is necessary.

In general I think it is unfortunate that we rely on invoking the 'junit' method explicitly in order for test failures to be reported. We should check out configuration of surefire/failsafe at some point.

mor-n4 commented:
@jpilgrim please delete the branch if you no longer need it. Thanks!
